src/Servers/Kestrel/Core/src/Internal/TlsConnectionFeature.cs

```diff
@@ -122,7 +129,19 @@ public int KeyExchangeStrength
 
     private async Task<X509Certificate2?> GetClientCertificateAsyncCore(CancellationToken cancellationToken)
     {
-        await _sslStream.NegotiateClientCertificateAsync(cancellationToken);
+        try
+        {
+            await _sslStream.NegotiateClientCertificateAsync(cancellationToken);
+        }
+        catch (Exception)
+        {
+            // We can't tell which exceptions are fatal or recoverable. Consider them all fatal
+            // and close the connection to avoid over-caching. This allows recovery by starting
+            // a new connection https://github.com/dotnet/aspnetcore/issues/41369
+            _context.Features.Get<IConnectionLifetimeNotificationFeature>()?.RequestClose();
```

Comment:
Why graceful?

Reply:
The app is still able to send a HTTP response for the current request in most cases. The issue the customer hit was the next request being affected by the cached error.

Reply:
Should add that as a comment.